frontend/app/controllers/Paypal.scala

```diff
@@ -0,0 +1,105 @@
+package controllers
+
+import play.api.mvc.{Controller, Action}
+import play.api.mvc.Results.BadRequest
+import play.api.libs.json.{Json, JsSuccess, JsError}
+import okhttp3.{OkHttpClient, FormBody, Request, Response}
+import com.netaporter.uri.Uri.parseQuery
+import configuration.Config
+
+object Paypal extends Controller {
+
+	// Payment token used to tie Paypal requests together.
+	case class Token (token: String)
+
+	// Json writers.
+	implicit val tokenWrites = Json.writes[Token]
+	implicit val tokenReads = Json.reads[Token]
+
+	// The parameters sent with every NVP request.
+	private val defaultNVPParams = Map(
+		"USER" -> Config.paypalUser,
+		"PWD" -> Config.paypalPassword,
+		"SIGNATURE" -> Config.paypalSignature,
+		"VERSION" -> Config.paypalNVPVersion)
+
+	// Takes a series of parameters, send a request to Paypal, returns response.
+	def nvpRequest (params: Map[String, String]) = {
+
+		val client = new OkHttpClient()
+		val reqBody = new FormBody.Builder()
+
+		for ((param, value) <- defaultNVPParams) reqBody.add(param, value)
+		for ((param, value) <- params) reqBody.add(param, value)
```

Comment:
This does look elegant, but it could also be written a tiny bit more succinctly:

```
defaultNVPParams.map(reqBody.add)
params.map(reqBody.add)
```

or even something like (I forget the real method name):

```
reqBody.addAll(defaultNVParams)
reqBody.addAll(params)
```